File: 0438.FindAllAnagramsinaString.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        ans = []
        count = [0] * 128
        for ch in p:
            count[ord(ch)] += 1
        left = right = 0
        c = n = len(p)
        while right < len(s):
            if count[ord(s[right])] > 0:
                c -= 1
            count[ord(s[right])] -= 1
            right += 1
            
            if c == 0:
                ans.append(left)
            
            if right - left == n:
                if count[ord(s[left])] >= 0:
                    c += 1
                count[ord(s[left])] += 1
                left += 1
        return ans
        

File: 0438.FindAllAnagramsinaString.go
func findAnagrams(s string, p string) []int {
    if len(s) < len(p) {
        return nil
    }
    cnt := make([]int, 26)
    for _, ch := range p {
        cnt[ch-'a']++
    }
    ans := []int{}
    count := len(p)
    left, right := 0, 0
    for right < len(s) {
        if cnt[s[right]-'a'] > 0 {
            count--
        }
        cnt[s[right]-'a']--
        right++
        if count == 0 {
            ans = append(ans, left)
        }
        if right-left == len(p) {
            if cnt[s[left]-'a'] >= 0 {
                count++
            }
            cnt[s[left]-'a']++
            left++
        }
    }
    return ans
}